File: src/main/java/com/kotakotik/chaoscraftfabric/chaos_event_stuff/ChaosEventRegister.java
package com.kotakotik.chaoscraftfabric.chaos_event_stuff;

import com.kotakotik.chaoscraftfabric.ChaoscraftFabric;
import com.kotakotik.chaoscraftfabric.utils.CCReflections;
import org.reflections.Reflections;

import javax.annotation.processing.RoundEnvironment;
import javax.lang.model.element.TypeElement;
import javax.lang.model.type.TypeMirror;
import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Type;
import java.util.*;

public class ChaosEventRegister {
    public static final List<Class<ChaosEvent>> eventClasses = new ArrayList<>();
    public static final List<ChaosEvent> events = new ArrayList<>();

    static {
        Set<Class<? extends ChaosEvent>> implem = CCReflections.getSubTypesOf(ChaosEvent.class);
        Set<Class<?>> annot = CCReflections.getTypesAnnotatedWith(ChaosEventReg.class);
        List<Class<?>> classes = new ArrayList<>(annot);
        classes.addAll(implem);
        for(Class<?> clazz : classes) {
            if(ChaosEvent.class.isAssignableFrom(clazz)) {
                eventClasses.add((Class<ChaosEvent>) clazz);
                try {
                    events.add((ChaosEvent) clazz.getConstructor().newInstance());
                } catch (InstantiationException | IllegalAccessException | InvocationTargetException | NoSuchMethodException e) {
                    e.printStackTrace();
                }
            } else {
                ChaoscraftFabric.log.warn(String.format("class %s is annoted with ChaosEventReg but does not implement ChaosEvent (%s)", clazz.getSimpleName(), clazz.getName()));
            }
        }
    }
}

File: src/main/kotlin/com/kotakotik/chaoscraft/chaos_event_stuff/ChaosEventRegister.kt
package com.kotakotik.chaoscraft.chaos_event_stuff

import com.kotakotik.chaoscraft.ChaoscraftKotlin
import com.kotakotik.chaoscraft.utils.CCReflections
import java.lang.reflect.InvocationTargetException

import java.util.ArrayList


object ChaosEventRegister {
    val eventClasses: MutableList<Class<ChaosEvent>> = ArrayList()
    val events: MutableList<ChaosEvent> = ArrayList()

    fun reg() {
        ChaoscraftKotlin.log.info("registering events",)
        val implem: Set<Class<out ChaosEvent>> = CCReflections.getSubTypesOf(ChaosEvent::class.java)
        val annot: Set<Class<*>> = CCReflections.getTypesAnnotatedWith(ChaosEventReg::class.java)
        val classes: MutableList<Class<*>> = ArrayList(annot)
        classes.addAll(implem)
        for (clazz in classes) {
            if (ChaosEvent::class.java.isAssignableFrom(clazz)) {
                eventClasses.add(clazz as Class<ChaosEvent>)
                var name: String = "?"
                try {
                    val ev: ChaosEvent = clazz.getConstructor().newInstance() as ChaosEvent
                    name = ev.getId()
                    events.add(ev)
                } catch (e: InstantiationException) {
                    e.printStackTrace()
                } catch (e: IllegalAccessException) {
                    e.printStackTrace()
                } catch (e: InvocationTargetException) {
                    e.printStackTrace()
                } catch (e: NoSuchMethodException) {
                    e.printStackTrace()
                }
                ChaoscraftKotlin.log.info("registered event ${name} (${clazz.name})")
            } else {
                ChaoscraftKotlin.log.warn("class ${clazz.simpleName} is annotated with ChaosEventReg but does not implement ChaosEvent (${clazz.name})")
            }
        }
        ChaoscraftKotlin.log.info("registered ${events.size} events",)
    }
}